### PR TITLE
DD-1532: more unit tests and fixes

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes DD-1532: more unit tests and fixes

# Description of changes

* third tranche of unit test
* Constructor of LayerManagerImpl checks for garbage in the stagingDir
* delete directories and file correctly from DB

# How to test

# Related PRs 
(Add links)
* 

# Notify
@DANS-KNAW/core-systems

### CODE REVIEW

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    @Test
+    public void should_create_the_staging_root() throws IOException {
+        assertThat(stagingDir).doesNotExist();
+        new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir));
+        assertThat(stagingDir).isEmptyDirectory();
+    }
+
+    @Test
+    public void should_use_the_existing_directory() throws IOException {
+        Files.createDirectories(stagingDir.resolve("1234567890123"));
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java]: Safer to use a local long type var and use that on both lines 38 and 42. It also makes it more explicit that this is the same value. Also modify this in the other tests.

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        var topLayerId = new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir))
+            .getTopLayer().getId();
+        assertThat(topLayerId).isEqualTo(1234567890123L);
+    }
+
+    @Test
+    public void should_throw_on_a_plain_file() throws IOException {
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java]:
```suggestion
    public void should_throw_on_a_regular_file_in_the_staging_root() throws IOException {
```

That is what it is mostly called. "Plain" could cause confusion with "plain text".

File: src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java [hunk 3]
```diff
@@ -40,37 +38,42 @@ public class LayerManagerImpl implements LayerManager {
...
         this(stagingRoot, archiveProvider, null);
     }
 
-    @SneakyThrows
-    private void initTopLayer() {
-        List<Path> paths;
-        try (Stream<Path> pathStream = Files.list(stagingRoot)) {
-            paths = pathStream.toList();
+    private void initTopLayer() throws IOException {
+        if (Files.notExists(stagingRoot)) {
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java:53]: Not sure about whether auto-creating the staging root can cause surprises, but I can't think of a concrete scenario, so I'll go along with your change.

File: src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java [hunk 4]
```diff
@@ -40,37 +38,42 @@ public class LayerManagerImpl implements LayerManager {
...
             topLayer = new LayerImpl(id, stagingRoot.resolve(Long.toString(id)), archiveProvider.createArchive(Long.toString(id)));
         }
     }
 
+    private String toValidLayerName(Path path) {
+        if (!path.toFile().isDirectory()) {
+            throw new IllegalStateException("Not a directory: " + path);
+        }
+        if(!path.getFileName().toString().matches("\\d{13,}")) {
+            // more than 13 digits in nov 2286
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java]: Our Vault is forever :)

File: src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java [hunk 5]
```diff
@@ -100,7 +103,11 @@ private void archive(Layer layer) {
 
     @Override
     public Layer getLayer(long id) {
-        if (stagingRoot.resolve(Long.toString(id)).toFile().exists() || archiveProvider.exists(Long.toString(id))) {
+        if (id == topLayer.getId()) {
+            // an empty layer would not be found in the staging root
```
Comment (janvanmansum) [inline src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java]: I don´t understand this comment. Wouldn´t there be an empty staging directory?

Returning the `topLayer` here is OK.

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Test
+    public void should_create_the_staging_root() throws IOException {
+        assertThat(stagingDir).doesNotExist();
+        new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir));
+        assertThat(stagingDir).isEmptyDirectory();
+    }
+
+    @Test
+    public void should_use_the_existing_directory() throws IOException {
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java]:
```suggestion
    public void should_use_the_existing_staging_directory() throws IOException {
```

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        Files.createDirectories(stagingDir);
+        Files.createFile(stagingDir.resolve("1234567890123"));
+
+        assertThatThrownBy ( () -> new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir)))
+            .isInstanceOf(IllegalStateException.class)
+            .hasMessage("Not a directory: target/test/LayerManagerConstructorTest/layer_staging/1234567890123");
+    }
+
+    @Test
+    public void should_throw_on_the_too_short_directory_name() throws IOException {
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java]:
```suggestion
    public void should_throw_on_staging_directory_name_that_is_too_short_to_be_a_recent_timestamp() throws IOException {
```

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        Files.createDirectories(stagingDir.resolve("123456789012"));
+
+        assertThatThrownBy ( () -> new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir)))
+            .isInstanceOf(IllegalStateException.class)
+            .hasMessage("Not a timestamp: target/test/LayerManagerConstructorTest/layer_staging/123456789012");
+    }
+
+    @Test
+    public void should_throw_on_an_alphanumeric_directory() throws IOException {
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java]:
```suggestion
    public void should_throw_on_staging_directory_name_that_contains_letters() throws IOException {
```

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        Files.createDirectories(stagingDir.resolve("abcdefghijklmnopqrstuvwxyz"));
+
+        assertThatThrownBy ( () -> new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir)))
+            .isInstanceOf(IllegalStateException.class)
+            .hasMessage("Not a timestamp: target/test/LayerManagerConstructorTest/layer_staging/abcdefghijklmnopqrstuvwxyz");
+    }
+
+    @Test
+    public void should_throw_on_a_directory_with_a_dot_in_the_name() throws IOException {
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java]:
```suggestion
    public void should_throw_on_staging_directory_name_that_contains_decimal_point() throws IOException {
```

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java [hunk 10]
```diff
@@ -15,5 +15,120 @@
...
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessageContaining("No layer found with id 1234567890123");
+
+    }
+
+    @Test
+    public void should_find_garbage_in_stagingDir_created_after_creating_LayerManagerImpl_object() throws IOException {
+        // Given
+        var layerManager = new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir));
+        assertThat(layerManager.getTopLayer().getId()).isNotEqualTo(0L);
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java]: Why this "assert"? Also, it is under the "Given" comment.

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java [hunk 11]
```diff
@@ -15,5 +15,120 @@
...
+        Files.createDirectories(stagingDir.resolve("0"));
+
+        //When
+        var layer = layerManager.getLayer(0L);
+
+        // Then
+        assertThat(layer.getId()).isNotEqualTo(layerManager.getTopLayer().getId());
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java:54]: I also don't understand the point of the test.

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java [hunk 12]
```diff
@@ -15,5 +15,120 @@
...
+    public void should_find_an_empty_archived_layer() throws Exception {
+        // Given
+        Files.createDirectories(archiveDir);
+        var layerManager = new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir));
+        var initialLayerId = layerManager.getTopLayer().getId();
+        assertThat(stagingDir).isEmptyDirectory(); // no content in the current layer
+        assertThat(archiveDir).isEmptyDirectory(); // nothing archived yet
+        sleep(1L); // to make sure to get a layer with another time stamp
+        layerManager.newTopLayer();
+        AssertionsForClassTypes.assertThat(initialLayerId).isNotEqualTo(layerManager.getTopLayer().getId());
```
Comment (janvanmansum) [inline src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java]: Use static import.

**Review by janvanmansum — CHANGES_REQUESTED** [review]
See inline comments. Mainly some i's to dot, but also one test that is unclear to me.

File: src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java (hunk 5 above)
Comment (jo-pol) [reply src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java]: A new layer in the staging directory is only created when something is added. Adjusted the comment.

File: src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java (hunk 11 above)
Comment (jo-pol) [reply src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java:54]: It shows inconsistent behavior between the constructor and getLayer.

**Review by janvanmansum — APPROVED** [review]

**Pull request merged**